councilmatic_core/migrations/0049_alter_person_headshot.py

```diff
@@ -0,0 +1,19 @@
+# Generated by Django 2.1.9 on 2019-07-17 15:16
+
+import django.core.files.storage
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('councilmatic_core', '0048_post_shape'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='person',
+            name='headshot',
+            field=models.FileField(default='images/headshot_placeholder.png', storage=django.core.files.storage.FileSystemStorage(base_url='/', location='/Users/goobzie/datamade/chi-councilmatic'), upload_to='images/headshots'),
```

Comment:
are there any negative side effects for this to be relative to your machine, or is it a placeholder?

Reply:
That's a really good question. I did some digging and it looks like there's already a migration that does this same operation, it just sets the default location to a path on _your_ machine:

https://github.com/datamade/django-councilmatic/blob/0cde8a0d90835d9918154d832dd1e38cc4596324/councilmatic_core/migrations/0046_subclass_ocd_billdocument.py#L30

I'm wondering if maybe new users will always generate new migrations since `location` is dynamically set in `models.py` based on the host environment:

https://github.com/datamade/django-councilmatic/blob/0cde8a0d90835d9918154d832dd1e38cc4596324/councilmatic_core/models.py#L20

I'll try to figure out if there's any way we can fix this.

Reply:
oh noooo lol

Reply:
looked into this. 

as best i can tell, the intent here was to store the headshots in the project static directory (that is, `<INSTANCE_APP>/static`) so they're collected and served as static files. this was advantageous because the default placeholder is a static file. by default, django file storage puts uploaded files in and serves them from the `MEDIA_ROOT` directory. having the uploaded headshots and the placeholder in two separate places – `STATIC_ROOT` v. `MEDIA_ROOT` – would preclude us from accessing their urls in a standard way in the templates, e.g., `{% static person.headshot.url %}` or simply `{% person.url %}`.

however, the current implementation programmatically builds an absolute path to the location of the static directory, which couples it to the structure of the developer's machine. surprisingly, i'm not sure that this be a problem because:

> All that will be stored in your database is a path to the file (relative to MEDIA_ROOT [or STATIC_PATH, in our implementation]). 

https://docs.djangoproject.com/en/2.2/ref/models/fields/#django.db.models.FileField.storage

django programmatically builds the full urls to files, based on the configured root directory.

still, it creates the annoying problem of django always thinking there are changes to the model on everyone's machine except for the last developer to create a migration setting the absolute path.

there are a few courses of action here:

1. **do nothing, and live with the migration snafu.** i don't like that this will train us to ignore messages about missing migrations from django. because we have relatively straightforward options for resolving this problem, i would not recommend this non-solution.
2. **implement a full storage subclass**. define a custom `__eq__` method that compares only the relative paths, to circumvent the absolute path problem: https://stackoverflow.com/a/32349636. this solution is quite clever. it would allow us to continue to use the static strategy and would spare us from having to update the template code to serve headshots from their `url`, without using the `static` method.

    on the other hand, this solution is quite clever. in a codebase with many clever things going on, i'd prefer not to introduce another piece of custom code to maintain/debug.
3. **use an absolute path for the default:** https://stackoverflow.com/a/47731766. django will only try to programmatically construct the url to a file if the filename is a relative path. otoh, it will simply return absolute paths as is, so we can set the default as a static file by providing an absolute path to the placeholder.

    this solution would require updates to the template code, so the app does not try to serve headshots from the static directory, however that behavior was introduced in the ocd model migration, so it will spare us from having to update template code in other instances when we upgrade them.

of these options, i would recommend number 3. i would also recommend doing that work in a separate pr, which i am happy to take on. what do you think, @jeancochrane (since this is your pr) and @fgregg (since you wrote the original code)?

Reply:
I also agree that #3 seems like a good candidate for a separate PR. Thanks for the great investigative work @hancush!

Reply:
could we just set the location in the migration to

`os.path.join(settings.STATIC_PATH)`

?